libs/core/algorithms/include/hpx/parallel/algorithms/fold.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/parallel/algorithms/reduce.hpp>
+
+namespace hpx::parallel { namespace detail {
+
+    template <typename _T>
+    struct fold_left : public algorithm<fold_left<_T>, _T>
+    {
+        constexpr fold_left() noexcept
+          : algorithm<fold_left, _T>("fold_left")
+        {
+        }
+
+        template <typename ExPolicy, typename FwdIter, typename Sent,
+            typename T, typename F>
+        HPX_HOST_DEVICE static constexpr T sequential(
+            ExPolicy&&, FwdIter first, Sent last, T&& init, F&& f)
+        {
+            hpx::reduce(first, last, HPX_FORWARD(T, init), HPX_FORWARD(F, f));
+        }
+
+        template <typename ExPolicy, typename FwdIter, typename Sent,
+            typename T, typename F>
+        static util::detail::algorithm_result_t<ExPolicy, FwdIter> parallel(
+            ExPolicy&& policy, FwdIter first, Sent last, T&& init, F&& f)
+        {
+            hpx::reduce(HPX_FORWARD(ExPolicy, policy), first, last, HPX_FORWARD(T, init),
+                HPX_FORWARD(F, f));
+        }
+    };
+
+}}    // namespace hpx::parallel::detail
+
+namespace hpx {
+inline constexpr struct fold_left_t final
+  : hpx::detail::tag_parallel_algorithm<fold_left_t>
+{
+private:
+    template <typename ExPolicy, typename FwdIter, typename T,
+        typename F>    // TODO : add concept
+    friend
+        typename hpx::parallel::util::detail::algorithm_result<ExPolicy>::type
+        tag_fallback_invoke(fold_left_t, ExPolicy&& policy, FwdIter first,
+            FwdIter last, T&& init, F&& f)
```

Comment:
The initial value `init` and the function `f` should be taken by value.

Reply:
Is there any reason we can't pass f by reference?

Reply:
a) the standard says to take a copy
b) this is to make sure that the lifetime of any state `F` may hold is decoupled from the invocation site